File: hammer/core/datahub/feature.py
from sqlalchemy import Column, Integer, String, UniqueConstraint
from sqlalchemy.dialects.postgresql import JSON

from ..base import Base


class Feature(Base):
    __tablename__ = "feature"

    name = Column(String(256), nullable=False, primary_key=True)
    version = Column(String(256), nullable=False, primary_key=True)
    source = Column(JSON, nullable=False)  # name, version, table_name
    entity = Column(JSON, nullable=False)  # name, version
    event_timestamp_field = Column(String(256))
    start_event_datetime = Column(String(64), default="", primary_key=True)
    end_event_datetime = Column(String(64), default="", primary_key=True)
    ttl = Column(Integer(), default=0)
    schema = Column(JSON, nullable=False)
    sink = Column(JSON, nullable=False)  # hive上 database, table
    transform = Column(String(1024), nullable=False)
    description = Column(String(256), default="")
    owner = Column(String(64), default="")
    status = Column(String(20), default="todo", nullable=False)

    __table_args__ = (UniqueConstraint("name", "version", "start_event_datetime", "end_event_datetime"),)

    def __repr__(self):
        return f"<Feature {self.name}:{self.version}>"

File: server/src/main/java/com/hammer/entity/Feature.java
package com.hammer.entity;

import java.util.Map;
import lombok.Data;


@Data
public class Feature {
    private String name;
    private String version;
    private Source source;
    private Entity entity;
    private String eventTimestampField;
    private String startEventDatetime;
    private String endEventDatetime;
    private int ttl;
    private Map<String, String> schema;  // JSON 结构
    private Map<String, String> sink;    // hive 上的 database, table
    private String transform;            // SQL 代码，传递给 FeatureComputer 进行计算
    private String description;
    private String owner;
    public String status; // 默认值为 "todo"

    public Feature(String name, String version, Source source, Entity entity,
                   String eventTimestampField, String startEventDatetime, String endEventDatetime, int ttl,
                   Map<String, String> schema, Map<String, String> sink, String transform, String description,
                   String owner, String status) {
        this.name = name;
        this.version = version;
        this.source = source;
        this.entity = entity;
        this.eventTimestampField = eventTimestampField;
        this.startEventDatetime = startEventDatetime;
        this.endEventDatetime = endEventDatetime;
        this.ttl = ttl;
        this.schema = schema;
        this.sink = sink;
        this.transform = transform;
        this.description = description;
        this.owner = owner;
        this.status = status;
    }
 
    @Override
    public String toString() {
        return String.format("<Feature %s:%s>", name, version);
    }
}

